File: app/src/main/java/com/babestudios/companyinfouk/ui/officerappointments/OfficerAppointmentsActivity.java
package com.babestudios.companyinfouk.ui.officerappointments;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.babestudios.companyinfouk.CompaniesHouseApplication;
import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.model.officers.appointments.Appointment;
import com.babestudios.companyinfouk.data.model.officers.appointments.Appointments;
import com.babestudios.companyinfouk.ui.company.CompanyActivity;
import com.babestudios.companyinfouk.uiplugins.BaseActivityPlugin;
import com.babestudios.companyinfouk.utils.DividerItemDecorationWithSubHeading;
import com.pascalwelsch.compositeandroid.activity.CompositeActivity;

import net.grandcentrix.thirtyinch.plugin.TiActivityPlugin;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;

public class OfficerAppointmentsActivity extends CompositeActivity implements OfficerAppointmentsActivityView,
		OfficerAppointmentsAdapter.AppointmentsRecyclerViewClickListener {

	@Bind(R.id.toolbar)
	Toolbar toolbar;

	@Bind(R.id.progressbar)
	ProgressBar progressbar;

	@Bind(R.id.officer_appointments_recycler_view)
	RecyclerView officerAppointmentsRecyclerView;

	@Inject
	OfficerAppointmentsPresenter officerAppointmentsPresenter;

	String officerId;

	TiActivityPlugin<OfficerAppointmentsPresenter, OfficerAppointmentsActivityView> officerAppointmentsActivityPlugin = new TiActivityPlugin<>(
			() -> {
				CompaniesHouseApplication.getInstance().getApplicationComponent().inject(this);
				return officerAppointmentsPresenter;
			});

	BaseActivityPlugin baseActivityPlugin = new BaseActivityPlugin();

	public OfficerAppointmentsActivity() {
		addPlugin(officerAppointmentsActivityPlugin);
		addPlugin(baseActivityPlugin);
	}

	@SuppressWarnings("ConstantConditions")
	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_officer_appointments);
		baseActivityPlugin.logScreenView(this.getLocalClassName());

		ButterKnife.bind(this);

		officerId = getIntent().getStringExtra("officerId");

		if (toolbar != null) {
			setSupportActionBar(toolbar);
			getSupportActionBar().setDisplayHomeAsUpEnabled(true);
			getSupportActionBar().setTitle(R.string.officer_appointments_title);
			toolbar.setNavigationOnClickListener(v -> onBackPressed());
		}

	}

	private void createChargesDetailsRecyclerView(Appointments appointments) {
		LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
		officerAppointmentsRecyclerView.setLayoutManager(linearLayoutManager);
		ArrayList<Integer> titlePositions = new ArrayList<>();
		titlePositions.add(0);
		officerAppointmentsRecyclerView.addItemDecoration(
				new DividerItemDecorationWithSubHeading(this, titlePositions));
		OfficerAppointmentsAdapter adapter = new OfficerAppointmentsAdapter(this, appointments);
		officerAppointmentsRecyclerView.setAdapter(adapter);
	}

	@Override
	public void showProgress() {
		progressbar.setVisibility(View.VISIBLE);
	}

	@Override
	public void hideProgress() {
		progressbar.setVisibility(View.GONE);
	}

	@Override
	public void showAppointments(Appointments appointments) {
		createChargesDetailsRecyclerView(appointments);
	}

	@Override
	public String getOfficerId() {
		return officerId;
	}

	@Override
	public void showError() {
		Toast.makeText(this, R.string.could_not_retrieve_officer_appointments_info, Toast.LENGTH_LONG).show();
	}

	@Override
	public void appointmentItemClicked(View v, int position, Appointment item) {
		startCompanyActivity(item.appointedTo.companyNumber, item.appointedTo.companyName);
	}

	@Override
	public void startCompanyActivity(String companyNumber, String companyName) {
		Intent startIntent = new Intent(this, CompanyActivity.class);
		startIntent.putExtra("companyNumber", companyNumber);
		startIntent.putExtra("companyName", companyName);
		baseActivityPlugin.startActivityWithRightSlide(startIntent);
	}

	@Override
	public void super_onBackPressed() {
		super.super_finish();
		super_overridePendingTransition(R.anim.left_slide_in, R.anim.left_slide_out);
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/officerappointments/OfficerAppointmentsActivity.kt
package com.babestudios.companyinfouk.ui.officerappointments

import android.content.Context
import android.content.Intent
import android.os.Bundle

import kotlinx.android.synthetic.main.activity_officer_appointments2.*
import com.babestudios.companyinfouk.R
import com.babestudios.base.mvp.ErrorType
import com.babestudios.base.mvp.list.BaseViewHolder
import com.babestudios.companyinfouk.ext.startActivityWithRightSlide
import com.uber.autodispose.AutoDispose
import com.uber.autodispose.ScopeProvider
import com.ubercab.autodispose.rxlifecycle.RxLifecycleInterop
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity
import io.reactivex.CompletableSource
import androidx.lifecycle.ViewModelProviders
import com.babestudios.base.view.MultiStateView.*
import com.babestudios.companyinfouk.Injector
import androidx.recyclerview.widget.LinearLayoutManager
import com.babestudios.base.view.DividerItemDecoration
import com.babestudios.base.view.EndlessRecyclerViewScrollListener
import com.babestudios.companyinfouk.ui.company.createCompanyIntent
import com.babestudios.companyinfouk.ui.officerappointments.list.*

import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.row_officer_appointments_header.*

private const val OFFICER_ID = "com.babestudios.companyinfouk.ui.officer_id"

class OfficerAppointmentsActivity : RxAppCompatActivity(), ScopeProvider {


	private var officerAppointmentsAdapter: OfficerAppointmentsAdapter? = null

	override fun requestScope(): CompletableSource = RxLifecycleInterop.from(this).requestScope()

	private val viewModel by lazy { ViewModelProviders.of(this).get(OfficerAppointmentsViewModel::class.java) }

	private lateinit var officerAppointmentsPresenter: OfficerAppointmentsPresenterContract

	private val eventDisposables: CompositeDisposable = CompositeDisposable()

	//region life cycle

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_officer_appointments2)
		setSupportActionBar(pabOfficerAppointments.getToolbar())
		supportActionBar?.setDisplayHomeAsUpEnabled(true)
		pabOfficerAppointments.setNavigationOnClickListener { onBackPressed() }
		supportActionBar?.setTitle(R.string.officer_appointments_title)
		when {
			viewModel.state.value.appointmentItems != null -> {
				initPresenter(viewModel)
			}
			savedInstanceState != null -> {
				savedInstanceState.getParcelable<OfficerAppointmentsState>("STATE")?.let {
					with(viewModel.state.value) {
						appointmentItems = it.appointmentItems
						officerId = it.officerId
					}
				}
				initPresenter(viewModel)
			}
			else -> {
				viewModel.state.value.officerId = intent.getStringExtra(OFFICER_ID)
				initPresenter(viewModel)
			}
		}

		createRecyclerView()
		observeState()
	}

	override fun onResume() {
		super.onResume()
		observeActions()
	}

	override fun onSaveInstanceState(outState: Bundle) {
		outState.putParcelable("STATE", viewModel.state.value)
		super.onSaveInstanceState(outState)
	}

	private fun initPresenter(viewModel: OfficerAppointmentsViewModel) {
		if (!::officerAppointmentsPresenter.isInitialized) {
			officerAppointmentsPresenter = Injector.get().officerAppointmentsPresenter()
			officerAppointmentsPresenter.setViewModel(viewModel, requestScope())
		}
	}

	private fun createRecyclerView() {
		val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
		rvOfficerAppointments?.layoutManager = linearLayoutManager
		rvOfficerAppointments.addItemDecoration(DividerItemDecoration(this))
		rvOfficerAppointments.addOnScrollListener(object : EndlessRecyclerViewScrollListener(linearLayoutManager) {
			override fun onLoadMore(page: Int, totalItemsCount: Int) {
				officerAppointmentsPresenter.loadMoreAppointments(page)
			}
		})
	}


	//endregion

	//region render

	private fun observeState() {
		viewModel.state
				.`as`(AutoDispose.autoDisposable(this))
				.subscribe { render(it) }
	}

	private fun render(state: OfficerAppointmentsState) {
		when {
			state.isLoading -> msvOfficerAppointments.viewState = VIEW_STATE_LOADING
			state.errorType != ErrorType.NONE -> msvOfficerAppointments.viewState = VIEW_STATE_ERROR
			state.appointmentItems == null -> msvOfficerAppointments.viewState = VIEW_STATE_EMPTY
			else -> {
				state.appointmentItems?.let {
					msvOfficerAppointments.viewState = VIEW_STATE_CONTENT
					textViewOfficerName.text = state.officerName
					if (rvOfficerAppointments?.adapter == null) {
						officerAppointmentsAdapter = OfficerAppointmentsAdapter(it, OfficerAppointmentsTypeFactory())
						rvOfficerAppointments?.adapter = officerAppointmentsAdapter
					} else {
						officerAppointmentsAdapter?.updateItems(it)
					}
					observeActions()
				}
			}
		}
	}

	//endregion

	//region events

	private fun observeActions() {
		eventDisposables.clear()
		officerAppointmentsAdapter?.getViewClickedObservable()
				?.take(1)
				?.`as`(AutoDispose.autoDisposable(this))
				?.subscribe { view: BaseViewHolder<AbstractOfficerAppointmentsVisitable> ->
					viewModel.state.value.appointmentItems?.let { appointmentItems ->
						val company = (appointmentItems[(view as OfficerAppointmentsViewHolder).adapterPosition] as OfficerAppointmentsVisitable).appointment.appointedTo
						company?.let { appointedTo ->
							startActivityWithRightSlide(this.createCompanyIntent(appointedTo.companyNumber!!, appointedTo.companyName!!))
						}
					}
				}
				?.let { eventDisposables.add(it) }
	}

	//endregion
}

fun Context.createOfficerAppointmentsIntent(officerId: String): Intent {
	return Intent(this, OfficerAppointmentsActivity::class.java)
			.putExtra(OFFICER_ID, officerId)
}
